Improve Visual Script editor to suggest the proper visual script nodes.

Gallilus opened the pull request:
# proposal [#3528](https://github.com/godotengine/godot-proposals/issues/3528)

While filters are preset they are fully editable while searching.
The base is set relative to what input the editor gives.

- It is way easier to use constants.
- Its is possible to find content, not inside the base.

![image](https://user-images.githubusercontent.com/20573784/147412844-bf924dec-9320-4d09-93ca-2b3c437e2fc3.png)
![image](https://user-images.githubusercontent.com/20573784/147413106-95785e7c-0d3a-4d7c-b7dc-aa3c3685a167.png)

https://user-images.githubusercontent.com/20573784/147594930-b8daf8bb-67a0-4fe2-90c8-9a33177e07d5.mp4

Fixes: https://github.com/godotengine/godot-proposals/issues/3528

Gallilus committed:
Improve Visual Script editor to suggest the proper visual script nodes.

Gallilus commented:
added a few tweaks and easier access to deconstructors.

fire reviewed (approved):
Approved on the design.

Need to build and check for crashes and mistakes on a technical review.

fire reviewed (commented):
Do the icons have tooltips?

Gallilus commented:
Yes, even the searchbar

fire reviewed (approved):
Looks good to me.

akien-mga commented:
Neither the current PR title nor the commit message ("Update visual script autocomplete") seem to be an accurate description of what this PR does, could you amend them?

See https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#format-your-commit-messages-with-readability-in-mind

fire commented:
@Gallilus What do you think about this phrasing?

>  Improve PropertySelector and EditorHelpSearch to have the capability contextually find the proper VSNodes.

Gallilus commented:
> 

Will amend tomorow.

fire commented:
https://user-images.githubusercontent.com/32321/148127391-3f87023e-5949-48ec-ad1e-624dec6c35bc.mp4

Material for a future or this pull request.

The ability to search for inputs.

Gallilus commented:
@akien-mga Happy new year.
And thanks for always having patience with my descriptions and writing.

fire commented on modules/visual_script/editor/visual_script_property_selector.h:
```diff
@@ -60,32 +102,118 @@ class VisualScriptPropertySelector : public ConfirmationDialog {
-	VBoxContainer *vbc;
-
-	Vector<Variant::Type> type_filter;
+	VBoxContainer *vbox;
```
Fix later. set to nullptr

fire replied:
Sorry if this was confusing. I mean that the variables should be intitialized to like nullptr or 0, but since there's like 15 variables to check it can be left to another commit.

fire reviewed (approved):
I concentrated testing on functionality so thanks for catching the commit message and also missed class variable inititializing.

fire commented:
@Chaosus Can you take a look?

Chaosus commented on modules/visual_script/editor/visual_script_property_selector.cpp:
```diff
@@ -55,679 +70,1200 @@ void VisualScriptPropertySelector::_sbox_input(const Ref<InputEvent> &p_ie) {
-		item->set_selectable(2, false);
-		item->set_metadata(2, connecting);
+void VisualScriptPropertySelector::select_from_visual_script(const Ref<Script> &p_script, bool clear_text) {
+	set_title(TTR("Select from visual script"));
```
Maybe the "Create Visual Script Node" title sounds better (just to be consistent with "Create Shader Node" in visual shaders).

Gallilus replied:
"Create Visual Script Node" is not accurate, it allows you to select a node compatible with the visual script you are editing.
In my mind, Visual script nodes are things like sequences or conditions similar to keywords in GDScript.

Chaosus commented:
Not sure whether it is related to this issue but there is a bunch of errors emitted when the node is created from base type(e.g Vector):

![vst_bug](https://user-images.githubusercontent.com/3036176/148733853-4bde6aec-30da-4153-9940-ec81848a6fb4.gif)

Chaosus commented:
When there is no text entered in the search text edit it shows only a few keywords:

![image](https://user-images.githubusercontent.com/3036176/148734032-4e713a38-a05e-4918-8726-adbb1a4c3b49.png)

Does this intended? It looks incorrect for me.

Chaosus commented:
Clicking on category emits errors:

![vst_bug2](https://user-images.githubusercontent.com/3036176/148734465-2e120f26-703a-4fc5-b482-095919e66fce.gif)

Add a check to prevent this from happening.

akien-mga commented:
> WIP for proposal #3528

Is it still WIP or ready? It would be good to amend the OP description if it's now fully implemented.

Gallilus commented:
Next week I will try to fix the errors

Gallilus commented:
Added the null pointers (28)

Fixed the error when dropping nodes.

When there is no text entered there are only a few keywords (visual script nodes) available. Yes else you are provided with too many options and there are no more options for flow control. To see all options press spacebar.

About the error when selecting greyed out options. Yes, it needs to be handled in further PR's. For the above example, it might be interesting to change the scope reference. What do you think people will want when they select a class?

I hope this is a mergeable starting point for further improvements.

Chaosus commented:
> When there is no text entered there are only a few keywords (visual script nodes) available. Yes else you are provided with too many options and there are no more options for flow control. To see all options press spacebar.

Alright, I'm not a specialist in visual scripts, so it might be ok. I just make a point that pressing the spacebar to show all members is a little bit counter-intuitive - in visual shaders, you see the full tree in the collapsed form at the beginning.

> About the error when selecting greyed out options. Yes, it needs to be handled in further PR's. For the above example, it might be interesting to change the scope reference. What do you think people will want when they select a class?

I don't know, maybe you should add a check to `ClassDB::is_class_exposed` to prevent this and allow instantiate a class, or add a special name setlist to check before pressing.

Nevertheless, I think it's in a good state to merge.

akien-mga commented:
Thanks!
